experiments/state/src/main/scala/org/apache/gearpump/streaming/state/api/CheckpointManager.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gearpump.streaming.state.api
+
+import org.apache.gearpump.TimeStamp
+
+/**
+ * CheckpointManager provides write/read user states to/from
+ * a persistent store (e.g. HDFS).
+ * It maintains
+ *   1. an internal clock that is synced with an reliable external source
+ *   2. a checkpoint time which is advanced continuously by the configured
+ *   checkpoint interval on each checkpoint commit
+ */
+trait CheckpointManager {
```

Comment:
each task would have its own checkpoint manager and perform checkpoint without interfering with others.